// jest-stare/lib/processor/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
const Constants_1 = require("./Constants");
const IJestStareConfig_1 = require("./doc/IJestStareConfig");
const EnvVars_1 = require("./EnvVars");
const IO_1 = require("../utils/IO");
class Config {
    constructor(mLogger, mExplicitConfig, mProcessParms) {
        this.mLogger = mLogger;
        this.mExplicitConfig = mExplicitConfig;
        this.mProcessParms = mProcessParms;
    }
    buildConfig() {
        const packageJsonConfig = this.getJestStareConfig();
        const envVars = new EnvVars_1.EnvVars();
        const mergedEnvAndPackageJsonConfig = envVars.resolve(packageJsonConfig, envVars.read());
        const config = this.mExplicitConfig || mergedEnvAndPackageJsonConfig;
        if (this.mExplicitConfig != null) {
            Object.keys(mergedEnvAndPackageJsonConfig).forEach((key) => {
                if (this.mExplicitConfig[key] == null && mergedEnvAndPackageJsonConfig[key] != null) {
                    config[key] = mergedEnvAndPackageJsonConfig[key];
                }
            });
        }
        if (config.resultDir == null) {
            config.resultDir = Constants_1.Constants.DEFAULT_RESULTS_DIR;
        }
        if (config.resultDir[config.resultDir.length - 1] !== "/") {
            config.resultDir += "/";
        }
        if (config.log != null) {
            this.mLogger.on = config.log;
        }
        if (this.mExplicitConfig != null) {
            if (this.mProcessParms && this.mProcessParms.reporter) {
            }
            else {
                this.mLogger.info(Constants_1.Constants.OVERRIDE_JEST_STARE_CONFIG);
            }
        }
        if (config.resultHtml == null) {
            this.mLogger.debug("Setting to default resultHtml");
            config.resultHtml = Constants_1.Constants.MAIN_HTML;
        }
        else {
            if (config.resultHtml.indexOf(Constants_1.Constants.HTML_EXTENSION) === -1) {
                config.resultHtml = config.resultHtml + Constants_1.Constants.HTML_EXTENSION;
            }
        }
        if (config.resultJson == null) {
            config.resultJson = Constants_1.Constants.RESULTS_RAW;
        }
        return config;
    }
    getJestStareConfig() {
        const packageJsonObject = IO_1.IO.readPackageJson();
        if (packageJsonObject[IJestStareConfig_1.PACKAGE_JSON_KEY] == null) {
            return {};
        }
        else {
            return packageJsonObject[IJestStareConfig_1.PACKAGE_JSON_KEY];
        }
    }
}
exports.Config = Config;
